{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "831211a0-0d7c-486b-941c-6ee9e07087d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f8121f4a-074e-4ac9-8843-f7787787f431",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/_y/kbhk52152hb2z95r0mn0x39w0000gr/T/ipykernel_61107/1090627024.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  Hs_df = pd.read_csv('/Users/susann/OBDS_course_Python/week_2/Homo_sapiens.GRCh38.107.gtf', sep=\"\\t\", header = None, comment=\"#\")   # read .tsv file into memory\n"
     ]
    }
   ],
   "source": [
    "Hs_df = pd.read_csv('/Users/susann/OBDS_course_Python/week_2/Homo_sapiens.GRCh38.107.gtf', sep=\"\\t\", header = None, comment=\"#\")   # read .tsv file into memory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e18e7e9d-cdaa-4bd8-9fc7-e8ac4b2979f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371239</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>32373</td>\n",
       "      <td>32528</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277475\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371240</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>31698</td>\n",
       "      <td>31841</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277475\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371241</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000275405\"; gene_version \"1\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371242</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000275405\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371243</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000275405\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3371244 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  0               1                2        3        4  5  6  \\\n",
       "0                 1  ensembl_havana             gene  1471765  1497848  .  +   \n",
       "1                 1  ensembl_havana       transcript  1471765  1497848  .  +   \n",
       "2                 1  ensembl_havana             exon  1471765  1472089  .  +   \n",
       "3                 1  ensembl_havana              CDS  1471885  1472089  .  +   \n",
       "4                 1  ensembl_havana      start_codon  1471885  1471887  .  +   \n",
       "...             ...             ...              ...      ...      ... .. ..   \n",
       "3371239  KI270713.1         ensembl   five_prime_utr    32373    32528  .  -   \n",
       "3371240  KI270713.1         ensembl  three_prime_utr    31698    31841  .  -   \n",
       "3371241  KI270713.1         ensembl             gene    21861    22024  .  -   \n",
       "3371242  KI270713.1         ensembl       transcript    21861    22024  .  -   \n",
       "3371243  KI270713.1         ensembl             exon    21861    22024  .  -   \n",
       "\n",
       "         7                                                  8  \n",
       "0        .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1        .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2        .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3        0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4        0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "...     ..                                                ...  \n",
       "3371239  .  gene_id \"ENSG00000277475\"; gene_version \"1\"; t...  \n",
       "3371240  .  gene_id \"ENSG00000277475\"; gene_version \"1\"; t...  \n",
       "3371241  .  gene_id \"ENSG00000275405\"; gene_version \"1\"; g...  \n",
       "3371242  .  gene_id \"ENSG00000275405\"; gene_version \"1\"; t...  \n",
       "3371243  .  gene_id \"ENSG00000275405\"; gene_version \"1\"; t...  \n",
       "\n",
       "[3371244 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Hs_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "449391db-be3c-4cf1-8d3d-55721c98b938",
   "metadata": {},
   "outputs": [],
   "source": [
    "Hs_df.columns=['seqname', 'sourc', 'mapped segments', 'unmapped segments']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
