{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "831211a0-0d7c-486b-941c-6ee9e07087d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f8121f4a-074e-4ac9-8843-f7787787f431",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/_y/kbhk52152hb2z95r0mn0x39w0000gr/T/ipykernel_61107/1090627024.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  Hs_df = pd.read_csv('/Users/susann/OBDS_course_Python/week_2/Homo_sapiens.GRCh38.107.gtf', sep=\"\\t\", header = None, comment=\"#\")   # read .tsv file into memory\n"
     ]
    }
   ],
   "source": [
    "Hs_df = pd.read_csv('/Users/susann/OBDS_course_Python/week_2/Homo_sapiens.GRCh38.107.gtf', sep=\"\\t\", header = None, comment=\"#\")   # read .tsv file into memory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e18e7e9d-cdaa-4bd8-9fc7-e8ac4b2979f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371239</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>32373</td>\n",
       "      <td>32528</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277475\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371240</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>31698</td>\n",
       "      <td>31841</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277475\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371241</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000275405\"; gene_version \"1\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371242</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000275405\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371243</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000275405\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3371244 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  0               1                2        3        4  5  6  \\\n",
       "0                 1  ensembl_havana             gene  1471765  1497848  .  +   \n",
       "1                 1  ensembl_havana       transcript  1471765  1497848  .  +   \n",
       "2                 1  ensembl_havana             exon  1471765  1472089  .  +   \n",
       "3                 1  ensembl_havana              CDS  1471885  1472089  .  +   \n",
       "4                 1  ensembl_havana      start_codon  1471885  1471887  .  +   \n",
       "...             ...             ...              ...      ...      ... .. ..   \n",
       "3371239  KI270713.1         ensembl   five_prime_utr    32373    32528  .  -   \n",
       "3371240  KI270713.1         ensembl  three_prime_utr    31698    31841  .  -   \n",
       "3371241  KI270713.1         ensembl             gene    21861    22024  .  -   \n",
       "3371242  KI270713.1         ensembl       transcript    21861    22024  .  -   \n",
       "3371243  KI270713.1         ensembl             exon    21861    22024  .  -   \n",
       "\n",
       "         7                                                  8  \n",
       "0        .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1        .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2        .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3        0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4        0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "...     ..                                                ...  \n",
       "3371239  .  gene_id \"ENSG00000277475\"; gene_version \"1\"; t...  \n",
       "3371240  .  gene_id \"ENSG00000277475\"; gene_version \"1\"; t...  \n",
       "3371241  .  gene_id \"ENSG00000275405\"; gene_version \"1\"; g...  \n",
       "3371242  .  gene_id \"ENSG00000275405\"; gene_version \"1\"; t...  \n",
       "3371243  .  gene_id \"ENSG00000275405\"; gene_version \"1\"; t...  \n",
       "\n",
       "[3371244 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Hs_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "449391db-be3c-4cf1-8d3d-55721c98b938",
   "metadata": {},
   "outputs": [],
   "source": [
    "Hs_df.columns=['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a86ca9ac-982e-430e-b0d5-50fcc8178c42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371239</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>32373</td>\n",
       "      <td>32528</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277475\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371240</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>31698</td>\n",
       "      <td>31841</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277475\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371241</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000275405\"; gene_version \"1\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371242</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000275405\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371243</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000275405\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3371244 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            seqname          source          feature    start      end score  \\\n",
       "0                 1  ensembl_havana             gene  1471765  1497848     .   \n",
       "1                 1  ensembl_havana       transcript  1471765  1497848     .   \n",
       "2                 1  ensembl_havana             exon  1471765  1472089     .   \n",
       "3                 1  ensembl_havana              CDS  1471885  1472089     .   \n",
       "4                 1  ensembl_havana      start_codon  1471885  1471887     .   \n",
       "...             ...             ...              ...      ...      ...   ...   \n",
       "3371239  KI270713.1         ensembl   five_prime_utr    32373    32528     .   \n",
       "3371240  KI270713.1         ensembl  three_prime_utr    31698    31841     .   \n",
       "3371241  KI270713.1         ensembl             gene    21861    22024     .   \n",
       "3371242  KI270713.1         ensembl       transcript    21861    22024     .   \n",
       "3371243  KI270713.1         ensembl             exon    21861    22024     .   \n",
       "\n",
       "        strand frame                                          attribute  \n",
       "0            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3            +     0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4            +     0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "...        ...   ...                                                ...  \n",
       "3371239      -     .  gene_id \"ENSG00000277475\"; gene_version \"1\"; t...  \n",
       "3371240      -     .  gene_id \"ENSG00000277475\"; gene_version \"1\"; t...  \n",
       "3371241      -     .  gene_id \"ENSG00000275405\"; gene_version \"1\"; g...  \n",
       "3371242      -     .  gene_id \"ENSG00000275405\"; gene_version \"1\"; t...  \n",
       "3371243      -     .  gene_id \"ENSG00000275405\"; gene_version \"1\"; t...  \n",
       "\n",
       "[3371244 rows x 9 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Hs_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "81899615-24fb-445c-9592-b83ccb4daf2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ensembl_havana', 'havana', 'ensembl', 'havana_tagene',\n",
       "       'ensembl_havana_tagene', 'mirbase', 'ensembl_tagene', 'insdc'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Hs_df['source'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3fe9f967-1c6a-4518-93cc-4b40f1cec99e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Hs_df['source'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "11623781-70cd-4e04-8130-f37e96ce668c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ens_df=Hs_df.query(\"source == 'ensembl_havana'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "08e796c8-de9c-4420-a92f-00e05c0150cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370044</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>46445284</td>\n",
       "      <td>46445769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370045</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>46445284</td>\n",
       "      <td>46445324</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370046</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>46445325</td>\n",
       "      <td>46445327</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370047</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>46324156</td>\n",
       "      <td>46324228</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370048</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>46445328</td>\n",
       "      <td>46445769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>714608 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source          feature     start       end score  \\\n",
       "0             1  ensembl_havana             gene   1471765   1497848     .   \n",
       "1             1  ensembl_havana       transcript   1471765   1497848     .   \n",
       "2             1  ensembl_havana             exon   1471765   1472089     .   \n",
       "3             1  ensembl_havana              CDS   1471885   1472089     .   \n",
       "4             1  ensembl_havana      start_codon   1471885   1471887     .   \n",
       "...         ...             ...              ...       ...       ...   ...   \n",
       "3370044      21  ensembl_havana             exon  46445284  46445769     .   \n",
       "3370045      21  ensembl_havana              CDS  46445284  46445324     .   \n",
       "3370046      21  ensembl_havana       stop_codon  46445325  46445327     .   \n",
       "3370047      21  ensembl_havana   five_prime_utr  46324156  46324228     .   \n",
       "3370048      21  ensembl_havana  three_prime_utr  46445328  46445769     .   \n",
       "\n",
       "        strand frame                                          attribute  \n",
       "0            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3            +     0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4            +     0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "...        ...   ...                                                ...  \n",
       "3370044      +     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...  \n",
       "3370045      +     2  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...  \n",
       "3370046      +     0  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...  \n",
       "3370047      +     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...  \n",
       "3370048      +     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...  \n",
       "\n",
       "[714608 rows x 9 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ens_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8c20f213-dfe2-4618-adb8-cc0bb70a3746",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'tuple' object is not callable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [15]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mens_df\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshape\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mTypeError\u001b[0m: 'tuple' object is not callable"
     ]
    }
   ],
   "source": [
    "ens_df.shape()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3d13342a-2da7-495b-9e2a-274a358e3482",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(714608, 9)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ens_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "340d33ff-6b7e-46c9-9ab3-a956780852c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pandas.core.groupby.generic.DataFrameGroupBy object at 0x7f872a2f2c70>"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Hs_df.groupby('feature')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "761171ec-ee8a-4ebb-89a2-811ee2f48e63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371239</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>32373</td>\n",
       "      <td>32528</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277475\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371240</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>31698</td>\n",
       "      <td>31841</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277475\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371241</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000275405\"; gene_version \"1\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371242</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000275405\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371243</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000275405\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3371244 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            seqname          source          feature    start      end score  \\\n",
       "0                 1  ensembl_havana             gene  1471765  1497848     .   \n",
       "1                 1  ensembl_havana       transcript  1471765  1497848     .   \n",
       "2                 1  ensembl_havana             exon  1471765  1472089     .   \n",
       "3                 1  ensembl_havana              CDS  1471885  1472089     .   \n",
       "4                 1  ensembl_havana      start_codon  1471885  1471887     .   \n",
       "...             ...             ...              ...      ...      ...   ...   \n",
       "3371239  KI270713.1         ensembl   five_prime_utr    32373    32528     .   \n",
       "3371240  KI270713.1         ensembl  three_prime_utr    31698    31841     .   \n",
       "3371241  KI270713.1         ensembl             gene    21861    22024     .   \n",
       "3371242  KI270713.1         ensembl       transcript    21861    22024     .   \n",
       "3371243  KI270713.1         ensembl             exon    21861    22024     .   \n",
       "\n",
       "        strand frame                                          attribute  \n",
       "0            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3            +     0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4            +     0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "...        ...   ...                                                ...  \n",
       "3371239      -     .  gene_id \"ENSG00000277475\"; gene_version \"1\"; t...  \n",
       "3371240      -     .  gene_id \"ENSG00000277475\"; gene_version \"1\"; t...  \n",
       "3371241      -     .  gene_id \"ENSG00000275405\"; gene_version \"1\"; g...  \n",
       "3371242      -     .  gene_id \"ENSG00000275405\"; gene_version \"1\"; t...  \n",
       "3371243      -     .  gene_id \"ENSG00000275405\"; gene_version \"1\"; t...  \n",
       "\n",
       "[3371244 rows x 9 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Hs_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1ae02302-4822-483d-b609-d923062b43c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "exon               1624585\n",
       "CDS                 871196\n",
       "transcript          251121\n",
       "three_prime_utr     203201\n",
       "five_prime_utr      171504\n",
       "start_codon          96934\n",
       "stop_codon           90724\n",
       "gene                 61860\n",
       "Selenocysteine         119\n",
       "Name: feature, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Hs_df[\"feature\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "bb26a526-8ca8-4879-9099-d654b40b496b",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_df=Hs_df.query(\"feature == 'gene'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "254c1e67-106f-4971-b7fd-9b922f1a8e55",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(61860, 9)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "7d283850-97e0-477d-b9eb-a5675576f2ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_df = genes_df.assign(gene_length = genes_df['end'] - genes_df['start'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "79a1b39f-7486-47fa-9631-cef8fa7af2f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>26083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2212523</td>\n",
       "      <td>2220738</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000234396\"; gene_version \"3\"; g...</td>\n",
       "      <td>8215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>629062</td>\n",
       "      <td>629433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225972\"; gene_version \"1\"; g...</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>8786211</td>\n",
       "      <td>8786913</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000224315\"; gene_version \"2\"; g...</td>\n",
       "      <td>702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>634376</td>\n",
       "      <td>634922</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198744\"; gene_version \"5\"; g...</td>\n",
       "      <td>546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371150</th>\n",
       "      <td>KI270711.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>4612</td>\n",
       "      <td>29626</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000271254\"; gene_version \"7\"; g...</td>\n",
       "      <td>25014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371224</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>30437</td>\n",
       "      <td>30580</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000275987\"; gene_version \"1\"; g...</td>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371227</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>35407</td>\n",
       "      <td>35916</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000268674\"; gene_version \"2\"; g...</td>\n",
       "      <td>509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371233</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>31698</td>\n",
       "      <td>32528</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277475\"; gene_version \"1\"; g...</td>\n",
       "      <td>830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371241</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000275405\"; gene_version \"1\"; g...</td>\n",
       "      <td>163</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61860 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            seqname          source feature    start      end score strand  \\\n",
       "0                 1  ensembl_havana    gene  1471765  1497848     .      +   \n",
       "107               1          havana    gene  2212523  2220738     .      +   \n",
       "111               1          havana    gene   629062   629433     .      +   \n",
       "114               1          havana    gene  8786211  8786913     .      -   \n",
       "117               1          havana    gene   634376   634922     .      +   \n",
       "...             ...             ...     ...      ...      ...   ...    ...   \n",
       "3371150  KI270711.1         ensembl    gene     4612    29626     .      -   \n",
       "3371224  KI270713.1         ensembl    gene    30437    30580     .      -   \n",
       "3371227  KI270713.1         ensembl    gene    35407    35916     .      +   \n",
       "3371233  KI270713.1         ensembl    gene    31698    32528     .      -   \n",
       "3371241  KI270713.1         ensembl    gene    21861    22024     .      -   \n",
       "\n",
       "        frame                                          attribute  gene_length  \n",
       "0           .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...        26083  \n",
       "107         .  gene_id \"ENSG00000234396\"; gene_version \"3\"; g...         8215  \n",
       "111         .  gene_id \"ENSG00000225972\"; gene_version \"1\"; g...          371  \n",
       "114         .  gene_id \"ENSG00000224315\"; gene_version \"2\"; g...          702  \n",
       "117         .  gene_id \"ENSG00000198744\"; gene_version \"5\"; g...          546  \n",
       "...       ...                                                ...          ...  \n",
       "3371150     .  gene_id \"ENSG00000271254\"; gene_version \"7\"; g...        25014  \n",
       "3371224     .  gene_id \"ENSG00000275987\"; gene_version \"1\"; g...          143  \n",
       "3371227     .  gene_id \"ENSG00000268674\"; gene_version \"2\"; g...          509  \n",
       "3371233     .  gene_id \"ENSG00000277475\"; gene_version \"1\"; g...          830  \n",
       "3371241     .  gene_id \"ENSG00000275405\"; gene_version \"1\"; g...          163  \n",
       "\n",
       "[61860 rows x 10 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "bc9edbd0-4dd4-4f50-862a-5915865b68d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32626.786744261233"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_df['gene_length'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4ae82953-254e-4ef9-97e0-ea69bf9f2945",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>6.186000e+04</td>\n",
       "      <td>6.186000e+04</td>\n",
       "      <td>6.186000e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>7.403611e+07</td>\n",
       "      <td>7.406874e+07</td>\n",
       "      <td>3.262679e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>5.558790e+07</td>\n",
       "      <td>5.559096e+07</td>\n",
       "      <td>9.230443e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>5.770000e+02</td>\n",
       "      <td>6.470000e+02</td>\n",
       "      <td>7.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>3.021300e+07</td>\n",
       "      <td>3.021955e+07</td>\n",
       "      <td>5.670000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>6.061628e+07</td>\n",
       "      <td>6.065367e+07</td>\n",
       "      <td>3.553500e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.092006e+08</td>\n",
       "      <td>1.092546e+08</td>\n",
       "      <td>2.519725e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2.489366e+08</td>\n",
       "      <td>2.489370e+08</td>\n",
       "      <td>2.473538e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              start           end   gene_length\n",
       "count  6.186000e+04  6.186000e+04  6.186000e+04\n",
       "mean   7.403611e+07  7.406874e+07  3.262679e+04\n",
       "std    5.558790e+07  5.559096e+07  9.230443e+04\n",
       "min    5.770000e+02  6.470000e+02  7.000000e+00\n",
       "25%    3.021300e+07  3.021955e+07  5.670000e+02\n",
       "50%    6.061628e+07  6.065367e+07  3.553500e+03\n",
       "75%    1.092006e+08  1.092546e+08  2.519725e+04\n",
       "max    2.489366e+08  2.489370e+08  2.473538e+06"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_df.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "9ad0c5b2-756c-4294-b273-5c20a520d2b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3553.5"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_df['gene_length'].median()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "3c6e752d-5356-459c-96e3-00c9e192b25a",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_sort_df = genes_df.sort_values(by=['gene_length'], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "1b0dffce-819c-4024-b7c3-3dfde8bfee25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2679474</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>5239802</td>\n",
       "      <td>7713340</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000078328\"; gene_version \"23\"; ...</td>\n",
       "      <td>2473538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1297140</th>\n",
       "      <td>7</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>146116002</td>\n",
       "      <td>148420998</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000174469\"; gene_version \"23\"; ...</td>\n",
       "      <td>2304996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1733179</th>\n",
       "      <td>9</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>8314246</td>\n",
       "      <td>10613002</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000153707\"; gene_version \"19\"; ...</td>\n",
       "      <td>2298756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1514121</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>31097677</td>\n",
       "      <td>33339609</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198947\"; gene_version \"18\"; ...</td>\n",
       "      <td>2241932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1850816</th>\n",
       "      <td>11</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>83455012</td>\n",
       "      <td>85627922</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000150672\"; gene_version \"18\"; ...</td>\n",
       "      <td>2172910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1601567</th>\n",
       "      <td>8</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2935353</td>\n",
       "      <td>4994972</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000183117\"; gene_version \"20\"; ...</td>\n",
       "      <td>2059619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3030998</th>\n",
       "      <td>20</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>13995369</td>\n",
       "      <td>16053197</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000172264\"; gene_version \"18\"; ...</td>\n",
       "      <td>2057828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1104549</th>\n",
       "      <td>6</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>63719980</td>\n",
       "      <td>65707226</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000188107\"; gene_version \"15\"; ...</td>\n",
       "      <td>1987246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338522</th>\n",
       "      <td>2</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>140231423</td>\n",
       "      <td>142131016</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000168702\"; gene_version \"18\"; ...</td>\n",
       "      <td>1899593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2041854</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>65912457</td>\n",
       "      <td>67763637</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000183230\"; gene_version \"18\"; ...</td>\n",
       "      <td>1851180</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature      start        end score strand  \\\n",
       "2679474      16  ensembl_havana    gene    5239802    7713340     .      +   \n",
       "1297140       7  ensembl_havana    gene  146116002  148420998     .      +   \n",
       "1733179       9  ensembl_havana    gene    8314246   10613002     .      -   \n",
       "1514121       X  ensembl_havana    gene   31097677   33339609     .      -   \n",
       "1850816      11  ensembl_havana    gene   83455012   85627922     .      -   \n",
       "1601567       8  ensembl_havana    gene    2935353    4994972     .      -   \n",
       "3030998      20  ensembl_havana    gene   13995369   16053197     .      +   \n",
       "1104549       6  ensembl_havana    gene   63719980   65707226     .      -   \n",
       "338522        2          havana    gene  140231423  142131016     .      -   \n",
       "2041854      10  ensembl_havana    gene   65912457   67763637     .      -   \n",
       "\n",
       "        frame                                          attribute  gene_length  \n",
       "2679474     .  gene_id \"ENSG00000078328\"; gene_version \"23\"; ...      2473538  \n",
       "1297140     .  gene_id \"ENSG00000174469\"; gene_version \"23\"; ...      2304996  \n",
       "1733179     .  gene_id \"ENSG00000153707\"; gene_version \"19\"; ...      2298756  \n",
       "1514121     .  gene_id \"ENSG00000198947\"; gene_version \"18\"; ...      2241932  \n",
       "1850816     .  gene_id \"ENSG00000150672\"; gene_version \"18\"; ...      2172910  \n",
       "1601567     .  gene_id \"ENSG00000183117\"; gene_version \"20\"; ...      2059619  \n",
       "3030998     .  gene_id \"ENSG00000172264\"; gene_version \"18\"; ...      2057828  \n",
       "1104549     .  gene_id \"ENSG00000188107\"; gene_version \"15\"; ...      1987246  \n",
       "338522      .  gene_id \"ENSG00000168702\"; gene_version \"18\"; ...      1899593  \n",
       "2041854     .  gene_id \"ENSG00000183230\"; gene_version \"18\"; ...      1851180  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_sort_df.iloc[0:10,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "d69c4f10-b494-435b-a91f-e5cd34a4b04b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "61860"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(genes_sort_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "c0cb096f-df9a-4897-a038-6b2ce4b9b301",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2418653</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105893542</td>\n",
       "      <td>105893561</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000211915\"; gene_version \"1\"; g...</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2489695</th>\n",
       "      <td>15</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>20004797</td>\n",
       "      <td>20004815</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000271317\"; gene_version \"1\"; g...</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2467325</th>\n",
       "      <td>15</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>21011451</td>\n",
       "      <td>21011469</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000270451\"; gene_version \"1\"; g...</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418624</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105884870</td>\n",
       "      <td>105884888</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227196\"; gene_version \"1\"; g...</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418616</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105881539</td>\n",
       "      <td>105881556</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225825\"; gene_version \"1\"; g...</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418705</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105910410</td>\n",
       "      <td>105910427</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000228131\"; gene_version \"1\"; g...</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418645</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105891191</td>\n",
       "      <td>105891207</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237020\"; gene_version \"1\"; g...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2467826</th>\n",
       "      <td>15</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>21017800</td>\n",
       "      <td>21017816</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000270185\"; gene_version \"1\"; g...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2462756</th>\n",
       "      <td>15</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>20011153</td>\n",
       "      <td>20011169</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000271336\"; gene_version \"1\"; g...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418701</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105909907</td>\n",
       "      <td>105909923</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237197\"; gene_version \"1\"; g...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418669</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105900638</td>\n",
       "      <td>105900654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227108\"; gene_version \"1\"; g...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418725</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105919502</td>\n",
       "      <td>105919518</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000236170\"; gene_version \"1\"; g...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418681</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105903616</td>\n",
       "      <td>105903631</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000232543\"; gene_version \"2\"; g...</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418657</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105894508</td>\n",
       "      <td>105894523</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227800\"; gene_version \"1\"; g...</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418713</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105913222</td>\n",
       "      <td>105913237</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000233655\"; gene_version \"1\"; g...</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2341080</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>22449113</td>\n",
       "      <td>22449125</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000228985\"; gene_version \"1\"; g...</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1312465</th>\n",
       "      <td>7</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>142786213</td>\n",
       "      <td>142786224</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000282431\"; gene_version \"1\"; g...</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418605</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105865551</td>\n",
       "      <td>105865561</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000236597\"; gene_version \"1\"; g...</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2341642</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>22439007</td>\n",
       "      <td>22439015</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237235\"; gene_version \"2\"; g...</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2341061</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>22438547</td>\n",
       "      <td>22438554</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223997\"; gene_version \"1\"; g...</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname  source feature      start        end score strand frame  \\\n",
       "2418653      14  havana    gene  105893542  105893561     .      -     .   \n",
       "2489695      15  havana    gene   20004797   20004815     .      -     .   \n",
       "2467325      15  havana    gene   21011451   21011469     .      -     .   \n",
       "2418624      14  havana    gene  105884870  105884888     .      -     .   \n",
       "2418616      14  havana    gene  105881539  105881556     .      -     .   \n",
       "2418705      14  havana    gene  105910410  105910427     .      -     .   \n",
       "2418645      14  havana    gene  105891191  105891207     .      -     .   \n",
       "2467826      15  havana    gene   21017800   21017816     .      -     .   \n",
       "2462756      15  havana    gene   20011153   20011169     .      -     .   \n",
       "2418701      14  havana    gene  105909907  105909923     .      -     .   \n",
       "2418669      14  havana    gene  105900638  105900654     .      -     .   \n",
       "2418725      14  havana    gene  105919502  105919518     .      -     .   \n",
       "2418681      14  havana    gene  105903616  105903631     .      -     .   \n",
       "2418657      14  havana    gene  105894508  105894523     .      -     .   \n",
       "2418713      14  havana    gene  105913222  105913237     .      -     .   \n",
       "2341080      14  havana    gene   22449113   22449125     .      +     .   \n",
       "1312465       7  havana    gene  142786213  142786224     .      +     .   \n",
       "2418605      14  havana    gene  105865551  105865561     .      -     .   \n",
       "2341642      14  havana    gene   22439007   22439015     .      +     .   \n",
       "2341061      14  havana    gene   22438547   22438554     .      +     .   \n",
       "\n",
       "                                                 attribute  gene_length  \n",
       "2418653  gene_id \"ENSG00000211915\"; gene_version \"1\"; g...           19  \n",
       "2489695  gene_id \"ENSG00000271317\"; gene_version \"1\"; g...           18  \n",
       "2467325  gene_id \"ENSG00000270451\"; gene_version \"1\"; g...           18  \n",
       "2418624  gene_id \"ENSG00000227196\"; gene_version \"1\"; g...           18  \n",
       "2418616  gene_id \"ENSG00000225825\"; gene_version \"1\"; g...           17  \n",
       "2418705  gene_id \"ENSG00000228131\"; gene_version \"1\"; g...           17  \n",
       "2418645  gene_id \"ENSG00000237020\"; gene_version \"1\"; g...           16  \n",
       "2467826  gene_id \"ENSG00000270185\"; gene_version \"1\"; g...           16  \n",
       "2462756  gene_id \"ENSG00000271336\"; gene_version \"1\"; g...           16  \n",
       "2418701  gene_id \"ENSG00000237197\"; gene_version \"1\"; g...           16  \n",
       "2418669  gene_id \"ENSG00000227108\"; gene_version \"1\"; g...           16  \n",
       "2418725  gene_id \"ENSG00000236170\"; gene_version \"1\"; g...           16  \n",
       "2418681  gene_id \"ENSG00000232543\"; gene_version \"2\"; g...           15  \n",
       "2418657  gene_id \"ENSG00000227800\"; gene_version \"1\"; g...           15  \n",
       "2418713  gene_id \"ENSG00000233655\"; gene_version \"1\"; g...           15  \n",
       "2341080  gene_id \"ENSG00000228985\"; gene_version \"1\"; g...           12  \n",
       "1312465  gene_id \"ENSG00000282431\"; gene_version \"1\"; g...           11  \n",
       "2418605  gene_id \"ENSG00000236597\"; gene_version \"1\"; g...           10  \n",
       "2341642  gene_id \"ENSG00000237235\"; gene_version \"2\"; g...            8  \n",
       "2341061  gene_id \"ENSG00000223997\"; gene_version \"1\"; g...            7  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_sort_df.iloc[len(genes_sort_df)-20:len(genes_sort_df),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84337831-1088-42a4-8616-04dd01793b40",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
